xul/bootstrap.js

```diff
@@ -0,0 +1,308 @@
+  it under the terms of the GNU Affero General Public License as published
+  by the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU Affero General Public License for more details.
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+let Cc = Components.classes, Ci = Components.interfaces, Cu = Components.utils;
+Cu.import("resource://gre/modules/Services.jsm");
+
+const branch = "extensions.esrc-explorer.";
+const apmo_pattern = /^https?:\/\/addons\.palemoon\.org\/addon\/(.*?)\//;
+const apmo_download_pattern = /^https?:\/\/addons\.palemoon\.org\/\?component=download&id=(.+?)&version=(.+?)&hash/;
+const apmo_match_pattern = '*://addons.palemoon.org/?component=download*';
+
+let styleSheetService = Cc["@mozilla.org/content/style-sheet-service;1"].getService(Ci.nsIStyleSheetService);
+let styleSheetURI = Services.io.newURI("chrome://esrc-explorer/skin/style.css", null, null);
+
+let gWindowListener, navigator;
+
+function ESrcExplorer(aWindow) {
+  this.init(aWindow);
+}
+ESrcExplorer.prototype = {
+
+  init: function(aWindow) {
+    this.browserWindow = aWindow;
+    this.tabBrowser = aWindow.gBrowser;
+
+    if (Services.prefs.getBoolPref(branch + "showButton")) {
+      this.addButton();
+    }
+
+    if (Services.prefs.getBoolPref(branch + "showContext")) {
+      this.addContext();
+    }
+
+    this.prefBranch = Services.prefs.getBranch(branch);
+    this.prefBranch.addObserver("", this, false);
+  },
+
+  done: function() {
+    this.prefBranch.removeObserver("", this);
+    this.prefBranch = null;
+
+    if (Services.prefs.getBoolPref(branch + "showButton")) {
+      this.removeButton();
+    }
+
+    if (Services.prefs.getBoolPref(branch + "showContext")) {
+      this.removeContext();
+    }
+
+    this.tabBrowser = null;
+    this.browserWindow = null;
+  },
+
+  observe: function(aSubject, aTopic, aData) {
+    if (aTopic != "nsPref:changed") return;
+    switch (aData) {
+      case 'showButton':
+        if (Services.prefs.getBoolPref(branch + "showButton")) {
+          this.addButton();
+        } else {
+          this.removeButton();
+        }
+        break;
+      case 'showContext':
+        if (Services.prefs.getBoolPref(branch + "showContext")) {
+          this.addContext();
+        } else {
+          this.removeContext();
+        }
+        break;
+    }
+  },
+
+  openExplorerOrSave: function(aUrl, aSave) {
+    let crx_url, filename;
+    if (apmo_pattern.test(aUrl)) {
+      crx_url = this.tabBrowser.getBrowserForTab(this.tabBrowser.selectedTab).
+                  contentDocument.getElementsByClassName("dllink_green")[0].href;
```

Comment:
This won't work with E10s, or when Palemoon's website changes, or when this is triggered from a context menu.

Is it possible to derive the download URL from the URL only?